### ISSUE TITLE
Add instances for `Reader`

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#6402]

**Comment by stephentetley** [comment]
We don't have a (binary) File type currently in the standard library to be test case for `Reader`. 

Should I make a wrapper over the JDK's `FileInputStream` as standard example / implementation?

**Comment by magnus-madsen** [comment]
> We don't have a (binary) File type currently in the standard library to be test case for `Reader`.
> 
> Should I make a wrapper over the JDK's `FileInputStream` as standard example / implementation?

I think so yes. I am not sure what we should do about the `File` module. It was intended as something quick and dirty, but I still think such a high-level API can be useful: sometimes you just want to dump strings into files other times you want precise control.

**Comment by magnus-madsen** [comment]
Perhaps `File` module should have been called "`Files`", i.e. its a collection of utility functions for working on files, not a file itself.

**Comment by stephentetley** [comment]
> Perhaps the `File` module should have been called "`Files`", i.e. its a collection of utility functions for working on files, not a file itself.

Yes, that seems to be the JDK way of doing things e.g. `java.nio.file.Files`

**Comment by magnus-madsen** [comment]
Note that Java's `File` / `Path` seems to be like a value object that just holds the name of a file whereas a Rust `File` is actually an open resource that you can read from:

https://doc.rust-lang.org/std/fs/struct.File.html

I understand you might want both (because paths are tricky), but I think it would be nice to have a Rust-style `File` that you can obtain from just a string (if you know you are in the "easy case") and then just read directly from.

**Comment by magnus-madsen** [comment]
Since Flix is not following Java terminology, perhaps we could have both `File`, `Path`, and `Files` where a `File` can be obtained from a string or via detour to `Path` (if one wants to be careful output legal file system paths/names).

**Comment by magnus-madsen** [comment]
For now, I think we should introduce a `File` which wraps a Java `FileInputStream` (or something to that effect).

**Comment by stephentetley** [comment]
I favour keeping the name `FileInputStream` because it's read only so only it only has half the capabilities what you might expect a "File" to have. When we get round to adding `FileOutputStream` that would have equal claim on the name `File` as it provides the write only view of things.

Should I pair `FileInputStream` with a region to emphasize it is an imperative "resource" like `StringBuilder` or is enough that all the operations working on it will use the `IO` effect?

i.e. 
~~~

enum FileInputStream[_: Region](##java.io.FileInputStream)

~~~

Or:

~~~

enum FileInputStream(##java.io.FileInputStream)

~~~

**Comment by magnus-madsen** [comment]
> I favour keeping the name FileInputStream because it's read only so only it only has half the capabilities what you might expect a "File" to have. When we get round to adding FileOutputStream that would have equal claim on the name File as it provides the write only view of things.

I would think that `File` would have both `Reader` and `Writer` instances, hence it could be used for both. This seems to be the case in Rust:

https://doc.rust-lang.org/std/fs/struct.File.html#impl-Read-for-File
https://doc.rust-lang.org/std/fs/struct.File.html#impl-Write-for-File

> Should I pair FileInputStream with a region to emphasize it is an imperative "resource" like StringBuilder or is enough that all the operations working on it will use the IO effect?

Regions are more like ST Monad which allows imperative functional programming. Reading and writing cannot be pure, so they should always be tagged with `IO`. 

A higher-level picture is this: We can write pure functions. We can functions that use IO. In the middle we will have algebraic effects, but ultimately the handlers for such effects will use IO (e.g. these Readers and Writers we are adding)-- at least if they want the effects to happen. Thus the whole setup will support functional programming with an imperative shell around it. The Readers and Writers are the imperative shell where the IO stuff actually happens. I hope that makes sense.

**Comment by stephentetley** [comment]
Wrapping `java.io.RandomAccessFile` looks like the way to go, unlike `FileInputStream` / `FileOutputStream` it has both read and write capabilities.

**Comment by stephentetley** [comment]
I've made PR #6422 to add a new `File` type.

`java.io.RandomAccessFile` turns out to be a bit disappointing (missing modern features - see comments in the PR) but `java.nio.channels.FileChannel` doesn't seem to be a suitable replacement.
